xds/internal/balancer/clustermanager/clustermanager.go

```diff
@@ -67,62 +69,107 @@ func (bb) ParseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, err
 }
 
 type bal struct {
-	logger *internalgrpclog.PrefixLogger
-
-	// TODO: make this package not dependent on xds specific code. Same as for
-	// weighted target balancer.
+	logger          *internalgrpclog.PrefixLogger
 	bg              *balancergroup.BalancerGroup
 	stateAggregator *balancerStateAggregator
 
 	children map[string]childConfig
 }
 
-func (b *bal) updateChildren(s balancer.ClientConnState, newConfig *lbConfig) {
+func (b *bal) setErrorPickerForChild(childName string, err error) {
+	b.logger.Warningf("%v", err)
+	b.stateAggregator.UpdateState(childName, balancer.State{
+		ConnectivityState: connectivity.TransientFailure,
+		Picker:            base.NewErrPicker(err),
+	})
+}
+
+func (b *bal) updateChildren(s balancer.ClientConnState, newConfig *lbConfig) error {
+	// TODO: Get rid of handling hierarchy in addresses. This LB policy never
+	// gets addresses from the resolver.
 	addressesSplit := hierarchy.Group(s.ResolverState.Addresses)
 
-	// Remove sub-pickers and sub-balancers that are not in the new cluster list.
+	// Remove sub-balancers that are not in the new list from the aggregator and
+	// balancergroup.
 	for name := range b.children {
 		if _, ok := newConfig.Children[name]; !ok {
 			b.stateAggregator.remove(name)
 			b.bg.Remove(name)
 		}
 	}
 
-	// For sub-balancers in the new cluster list,
-	// - add to balancer group if it's new,
-	// - forward the address/balancer config update.
-	for name, newT := range newConfig.Children {
-		if _, ok := b.children[name]; !ok {
-			// If this is a new sub-balancer, add it to the picker map.
-			b.stateAggregator.add(name)
-			// Then add to the balancer group.
-			b.bg.Add(name, balancer.Get(newT.ChildPolicy.Name))
+	var retErr error
+	for childName, childCfg := range newConfig.Children {
+		if _, ok := b.children[childName]; !ok {
+			// Add new sub-balancers to the aggregator and balancergroup.
+			b.stateAggregator.add(childName)
+			b.bg.Add(childName, balancer.Get(childCfg.ChildPolicy.Name))
 		} else {
-			// Already present, check for type change and if so send down a new
-			// builder.
-			if newT.ChildPolicy.Name != b.children[name].ChildPolicy.Name {
-				// Safe to ignore the returned error value because ParseConfig
-				// ensures that the child policy name is a registered LB policy.
-				b.bg.UpdateBuilder(name, newT.ChildPolicy.Name)
+			// If the child policy type has changed for existing sub-balancers,
```

Comment:
If the name didn't change, don't we still need to call UpdateCCS in case the config changed?  Or is this an optimization that the name always changes if the config changes?

Reply:
It was being taken care of in the call to `UpdateClientConnState` outside of the `if-else`. But I restructured this block a little so that we have a single call to `UpdateClientConnState` that will cover all cases.